integration-cli/graph_deadlock_test.go

```diff
@@ -0,0 +1,76 @@
+package main
+
+import (
+	"fmt"
+
+	"github.com/go-check/check"
+)
+
+// To reproduce the deadlock we have to create many images before the test.
+func createImages(c *check.C, n int) {
+	for i := 0; i < n; i++ {
+		dockerfile := fmt.Sprintf(`
+	FROM busybox
+	RUN echo a >/t%d.txt
+	RUN echo b >/%d.txt`, i, i)
+		_, err := buildImage("dummy", dockerfile, false)
+		c.Assert(err, check.IsNil)
+	}
+}
+
+func (s *DockerSuite) TestBuildDeadlock(c *check.C) {
+	createImages(c, 50)
+	dockerfile := `
+	FROM busybox
+	RUN echo a >/1.txt
+	RUN echo b >/2.txt
+	RUN echo c >/3.txt
+	RUN ps -ef >/ps1.txt
+	RUN ps -ef >/ps2.txt
+	RUN ps -ef >/ps3.txt
+	RUN ps -ef >/ps4.txt
+	RUN ps -ef >/ps5.txt
+	RUN ps -ef >/ps6.txt
+	`
+	ch := make(chan struct{})
+	var err error
+	go func() {
+		_, err = buildImage("dummy", dockerfile, false)
+		close(ch)
+	}()
+
+	for i := 0; i < 100; i++ {
+		dockerCmd(c, "images")
+	}
+	<-ch
+	c.Assert(err, check.IsNil)
+}
+
+func (s *DockerSuite) TestRmiDeadlock(c *check.C) {
+	createImages(c, 50)
+	dockerfile := `
+	FROM busybox
+	RUN echo a >/1.txt
+	RUN echo b >/2.txt
+	RUN echo c >/3.txt
+	RUN ps -ef >/ps1.txt
+	RUN ps -ef >/ps2.txt
+	RUN ps -ef >/ps3.txt
+	RUN ps -ef >/ps4.txt
+	RUN ps -ef >/ps5.txt
+	RUN ps -ef >/ps6.txt
+	`
+	id, err := buildImage("dummy", dockerfile, false)
+	c.Assert(err, check.IsNil)
+
+	ch := make(chan struct{})
+	go func() {
+		dockerCmd(c, "rmi", id)
+		close(ch)
+	}()
+
+	for i := 0; i < 100; i++ {
+		dockerCmd(c, "images")
+	}
+	<-ch
```

Comment:
It's just making sure that the `rmi` is complete before moving on, equivalent to

```
wg.Add(1)
go func() {
  rmi
  wg.Done()
}()

for things { stuff }
wg.Wait()
```

Otherwise if the `rmi` deadlocked the function would return anyway and not care.